src/utils/contexts/CartContext.jsx

```diff
@@ -0,0 +1,51 @@
+import React, { useState, useContext } from "react";
+
+const CartContext = React.createContext();
+
+const CartContextProvider = ({ children }) => {
+  const [cartItems, setCartItems] = useState([]);
+
+  const addProduct = (item) => {
+    const existing = cartItems.find((f) => f.product.id === item.product.id);
+    if (existing) {
+      existing.quantity += item.quantity;
+      const newList = [
+        ...cartItems.filter((f) => f.product.id !== item.product.id),
+        existing,
+      ];
+      setCartItems(newList);
+    } else {
+      setCartItems([...cartItems, item]);
+    }
+  };
+
+  const updateProduct = (quantity, id) => {
+    const existing = cartItems.find(({ product }) => product.id === id);
+    debugger
+    if (quantity === '' || (quantity > 0  && existing.product.data.stock >= quantity)) {
+      existing.quantity = quantity;
+      const newList = [
+        ...cartItems.filter(({ product }) => product.id !== id),
+        existing,
+      ];
+      setCartItems(newList);
+    }
+  };
```

Comment:
this function is updating an item in the array but the item is being mutated. Immutability is a good practice in JS, this code can something like:

```suggestion
  const updateProduct = (quantity, id) => {
    const existing = cartItems.find(({ product }) => product.id === id);

    if (quantity === '' || (quantity > 0 && existing.product.data.stock >= quantity)) {
      const updatedItem = {
        ...existing,
        quantity,
      };
      const newList = [
        ...cartItems.filter(({ product }) => product.id !== id),
        updatedItem,
      ];
      setCartItems(newList);
    }
  };
```